Tests/NIOPosixTests/DatagramChannelTests.swift

```diff
@@ -698,30 +698,37 @@ final class DatagramChannelTests: XCTestCase {
                 XCTAssertNoThrow(try sendChannel.close().wait())
             }
             
-            var buffer = sendChannel.allocator.buffer(capacity: 1)
-            buffer.writeRepeatingByte(0, count: 1)
             let ecnStates: [NIOExplicitCongestionNotificationState] = [.transportNotCapable,
                                                                        .congestionExperienced,
                                                                        .transportCapableFlag0,
                                                                        .transportCapableFlag1]
-            for ecnState in ecnStates {
-                let writeData = AddressedEnvelope(remoteAddress: receiveChannel.localAddress!,
-                                                  data: buffer,
-                                                  metadata: .init(ecnState: ecnState, packetInfo: expectedPacketInfo))
+            for (ecnStateIdx, ecnState) in ecnStates.enumerated() {
+                // Datagrams may be received out-of-order, so we use the payload to associate the read.
+                let writeData = AddressedEnvelope(
+                    remoteAddress: receiveChannel.localAddress!,
+                    data: sendChannel.allocator.buffer(integer: ecnStateIdx),
+                    metadata: .init(ecnState: ecnState, packetInfo: expectedPacketInfo)
+                )
                 // Sending extra data without flushing should trigger a vector send.
                 if (vectorSend) {
                     sendChannel.write(writeData, promise: nil)
                 }
                 try sendChannel.writeAndFlush(writeData).wait()
             }
 
-            let expectedReads = ecnStates.count * (vectorSend ? 2 : 1)
-            let reads = try receiveChannel.waitForDatagrams(count: expectedReads)
-            XCTAssertEqual(reads.count, expectedReads)
-            for readNumber in 0..<reads.count {
-                let read = reads[readNumber]
-                XCTAssertEqual(read.metadata?.ecnState, ecnStates[readNumber / (vectorSend ? 2 : 1)])
-                XCTAssertEqual(read.metadata?.packetInfo, expectedPacketInfo)
+            let expectedNumReads = ecnStates.count * (vectorSend ? 2 : 1)
+            let reads = try receiveChannel.waitForDatagrams(count: expectedNumReads)
+            XCTAssertEqual(reads.count, expectedNumReads)
+            for (ecnStateIdx, ecnState) in ecnStates.enumerated() {
+                // Datagrams may be received out-of-order, let's find the associated reads.
+                let encStateReads = reads.filter { read in
```

Comment:
Rather than this, can I suggest that we `map` the `reads` into an `Array` of `(ecnStateIdx, metadata)` and then sort it by `ecnStateIdx`? We should then be able to just compare for equality.

Reply:
We'll then still need to check that each `ecnStateIdx` is seen the expected number of times (depending on `vectorSend`) though? I guess I could build a dictionary of type `[ecnStateIdx: [metadata]]` and check that all the values have the expected length but I'm concerned that this would read less clearly than the code below.

Is the concern the performance off this test or the readability?

Reply:
We shouldn't need to check the number of times: as `ecnStateIdx` is in the array, the equality check does that for us. My concern is principally the readability: I think as written it is not clear that this test is attempting to confirm that we got the same data going in and coming out. The easiest way to confirm that is to take the reads and transform them into the same state we sent them in.

Reply:
👌 How about 5122d9efb, then?